src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2200,22 +2201,30 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
 			maxShardOffset = -1;
 		}
 
+		/*
+		 * For left joins we don't care about the shards pruned for the right hand side.
+		 * If the right hand side would prune to a smaller set we should still send it to
+		 * all tables of the left hand side. However if the right hand side is bigger than
+		 * the left hand side we don't have to send the query to any shard that is not
+		 * matching anything on the left hand side.
+		 *
+		 * Instead we will simply skip any RelationRestriction if it is an OUTER join and
+		 * the table is part of the non-outer side of the join.
+		 */
+		if (IsNonOuterTableOfOuterJoin(relationRestriction))
+		{
+			continue;
```

Comment:
Could this instead be written as a function without the negated naming and without `continue` jump. Possibly redusing mental load as:
```
if(IsOuterTableOfOuterJoin(...))
{
    foreach...
}
```

Reply:
I am not happy with the naming, it kind of grew like this while I was finding the correct logic here.

I don't know if moving the foreach into the if over the continue will make the whole block easier to read though. Some lines above this there is already a reference table check with a continue in the loop.

Initially I had this continue condition grouped with the reference table check but that failed our test suite due to the somewhat late check of shard count equality in the planning cycle.

re the regression test comment above; I will definitely add adequate tests for this to prevent further regressions, and making sure the condition on the join clauses is a correct one.